### ISSUE TITLE
Additional properties on CardElements and actions

### ISSUE THREAD

**Issue by matthidinger** [microsoft/AdaptiveCards#234]
# Intro

The properties that exist today on Cards are **official** properties and are **universally supported**. For example, `TextBlock.Weight` is supported and must be implemented by every renderer.

Some hosts want to add **host-specific properties** that they understand. This might be because their host supports advanced scenarios, like blurring an image or modifying the Action.Submit to perform a background activation instead of a foreground activation.

These **host-specific properties** are, by definition, **not universally supported**.

We should consider possible collisions and provide guidelines to hosts, as well as be mindful of the fact that an extension property in v1 may become part of the standard in v2, potentially with a different schema.

## Option 1
Seen below, **blur** being the extra, non-standard property.

```json
{
   "type": "Image",
   "url": "http://",
   "-ms-blur": true
}
```

## Option 2

```json
{
    "type": "Image",
    "url": "http://",
    "additionalProperties": [
        { "name": "-ms-blur", "value": "true" }
    ]
}
```

## Option 3

```json
{
    "type": "Image",
    "url": "http://",
    "additionalProperties": { 
             "Microsoft": { "blur": true }
        }
}
```

# Property name format

## Option 1 - Explicit to the host

```json
{
   "type": "Image",
   "url": "http://",
   "-ms-blur": true
}
```

## Option 2 - Generic

```json
{
   "type": "Image",
   "url": "http://",
   "-blur": true
}
```

The risk with option 2 is that if two different hosts decide to add `-blur`, they might conflict if they use different values, for example, one host might decide their blur is like the following...

```json
{
   "type": "Image",
   "url": "http://",
   "-blur": {
      "amount": "0.7",
      "color": "#FF00FF"
   }
}
```

This would break on the first host that expects a Boolean. Additional properties are, by definition, not formally accepted, and therefore **the schema of their values is not universally shared**. If the additional properties were formally accepted, we would simply make them official properties! **Therefore, not having a vendor prefix is not a valid option**. There needs to be some sort of prefix that scopes it correctly.

Additionally, if it was simply `-blur`, that begs the question of why not just make it a normal property, `blur`? If there's no vendor/product prefix, that implies that this property would be universally supported, just like `url` and `size` are.

## Prefix to the company? To the product name?

The next decision is how specific the prefix should be. Should it simply be the company name, `-ms-blur`? Or should it be `-ms-toast-blur` and `-ms-timeline-blur`?

As we saw above, the prefix serves two purposes...

1. Prevents conflicts from two separate hosts that didn't mutually agree on the schema
2. Clarifies that the property is specific to a host

With official properties, our rules are that officially properties are universally supported.

Therefore, the prefix clarifies that the property is only supported by the prefixed-renderer.

- If all Microsoft products universally support their definition of `blur`
  - Then `-ms-blur` makes sense. When using the property, by the properties name, you know where it WILL work (Microsoft products) and where it WON'T work (Facebook products, etc)
- If only one or a few Microsoft products support `blur`
  - Then `-ms-toast-blur` makes sense, targeting it to the specific product. You then know with confidence where the property will and won't work

# Retrieval from the Host app:

Here's one potential way we can solve it in the .NET renderer. Open to other suggestions. We want to support complex types.

```csharp
JObject blurObj = image.AdditionalProperties.Get("-ms-blur");
string blurStr = blurObj.GetAsString();
```

**Comment by dclaux** [comment]
The more I think about this the less I believe it is necessary to formalize such a mechanism, especially if we're going to allow a random object. I think all hosts should just be able to define whichever custom properties they want on any Adaptive Element type, including the card itself.

**Comment by Assaya** [comment]
It's a good idea if all web applications are compliable, happen Microsoft
SHOULD lay down some stringent measures.
Angela

**Comment by dclaux** [comment]
I do not see a need to have standard parsing of non-standard properties given that by definition there can also not be any standard use of such properties; hosts will have to implement custom code to consume them. If that's the case, why not let hosts implement the parsing of these extra properties using a renderer's extensibility model.

Unless renderers are able to consume the extra properties somehow (maybe if/when we introduce built-in data binding), there is not much value add in this feature IMO.

**Comment by dclaux** [comment]
@rihei until recently there was something missing in the JS SDK. It was already possible to parse additional properties on elements using the AdaptiveCard.onParseElement event, but the same wasn't available for actions. onParseAction was however introduced in the last couple months, and that would be the right way to do what you want without having to subclass OpenUrlAction.

Sorry for this oversight.

**Comment by rihei** [comment]
@dclaux Ok, thank you for the information! It seems that this onParseAction is not yet published to npm?

**Comment by andrewleader** [comment]
Closing as we implemented the `-ms-customProperty` style back in 1.0/1.1

**Comment by Assaya** [comment]
https://open.spotify.com/track/18QOV0BXTC7Z6TcQwORwDq?si=5UdgGpRVSDG0LhB9q38jKQ


As long as the moon can keep on shining...





angela@thehelm

[cid:1b7caa5e-7ab8-4762-95e3-b2825a076b89]
